### PR TITLE
[CLI-2642] Resolve race condition resulting in double channel close panic in `topic produce` commands

### PR DESCRIPTION
**Pull request by kevin-wu24**
Release Notes
-------------
Bug Fixes
- Resolved a close of closed channel panic in `topic produce`.

Checklist
---------
1. [CRUCIAL] Is the change for features that are already live in prod?
   * yes: ok

What
----
Building off of the ideas in https://github.com/confluentinc/cli/pull/2199, this race condition seems to have been resolved. 

Basically, sending an interrupt signal via `^C` now writes `\u0004` (Unicode encoding of `^D`) to the `input` channel, which can normally never receive this value, since writing `^D` to `stdIn` will just close the channel. 

Now, in the loop over `input`, we `break` if we get an error from the Scanner or if the value received from the `input` channel is `^D` (meaning a shutdown has been triggered with `^C`). This means the `input` channel is only closed once, when the user explicitly types `^D` into `stdIn`. 

References
----------
https://confluentinc.atlassian.net/browse/CLI-2642

Test & Review
-------------
Passes `make test` and manual testing. So far I haven't been able to recreate the `panic` and `topic produce` otherwise works as intended.

### CODE REVIEW

File: internal/kafka/command_topic.go [hunk 1]
```diff
@@ -15,7 +15,11 @@ import (
 	"github.com/confluentinc/cli/v3/pkg/log"
 )
 
-const numPartitionsKey = "num.partitions"
+const (
+	// EOF Unicode encoding for Ctrl+D (^D) character
+	EOF              = "\u0004"
```
Comment (brianstrauch) [inline internal/kafka/command_topic.go:20]: Based on your description, it doesn't matter what this value is equal to. Let's just make it "EOF"?

File: internal/kafka/command_topic.go [hunk 2]
```diff
@@ -15,7 +15,11 @@ import (
...
+	numPartitionsKey = "num.partitions"
```
Comment (brianstrauch) [inline internal/kafka/command_topic.go:21]: These constants are unrelated and shouldn't be grouped together.

File: internal/kafka/command_topic_produce.go [hunk 3]
```diff
@@ -132,20 +132,21 @@ func (c *command) produce(cmd *cobra.Command, args []string) error {
...
-			close(input)
-		}
+		input <- EOF
 	}()
 	// Prime reader
 	go scan()
 
 	deliveryChan := make(chan ckafka.Event)
 	for data := range input {
 		if data == "" {
```
Comment (brianstrauch) [inline internal/kafka/command_topic_produce.go:142]: `switch data`?

File: internal/kafka/command_topic_produce_onprem.go [hunk 4]
```diff
@@ -171,15 +171,20 @@ func (c *command) produceOnPrem(cmd *cobra.Command, args []string) error {
...
 
 	deliveryChan := make(chan ckafka.Event)
 	for data := range input {
 		if data == "" {
+			if scanErr != nil {
+				break
+			}
 			go scan()
 			continue
+		} else if data == EOF {
```
Comment (brianstrauch) [inline internal/kafka/command_topic_produce_onprem.go:186]: I wonder if there's a good way to combine the cloud/on-prem implementations? It's annoying that we have to keep updating both.

File: internal/kafka/command_topic.go (hunk 1 above)
Comment (kevin-wu24) [reply internal/kafka/command_topic.go:20]: If I make it "EOF", typing "EOF" into `stdIn` will `break` out of the loop, since we're checking the value of `data`. I actually tried exactly this haha!

Comment (brianstrauch) [reply internal/kafka/command_topic.go:20]: Ah, got it. Does pressing Ctrl-C still print "^C" to the terminal?

Comment (kevin-wu24) [reply internal/kafka/command_topic.go:20]: yup!

File: internal/kafka/command_topic_produce_onprem.go (hunk 4 above)
Comment (kevin-wu24) [reply internal/kafka/command_topic_produce_onprem.go:186]: Yeah, I think that would be good, but it might be better to put in another PR, since there's other `kafka` and even `kafka topic` commands that whose cloud/on-prem implementations share a lot of code, and that would end up being a lot of code refactoring.

File: internal/kafka/command_topic_produce.go (hunk 3 above)
Comment (kevin-wu24) [reply internal/kafka/command_topic_produce.go:142]: I don't think we can `switch data` because we need to `break` out of the loop over `input`.

**Review by brianstrauch — APPROVED** [review]
nice job!!
